peer/peer.go

```diff
@@ -0,0 +1,2013 @@
+
+	// User agent string to be used in peer messages.
+	UserAgentName string
+
+	// User agent version to be used in peer messages.
+	UserAgentVersion string
+}
+
+// minUint32 is a helper function to return the minimum of two uint32s.
+// This avoids a math import and the need to cast to floats.
+func minUint32(a, b uint32) uint32 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// newNetAddress attempts to extract the IP address and port from the passed
+// net.Addr interface and create a bitcoin NetAddress structure using that
+// information.
+func newNetAddress(addr net.Addr, services wire.ServiceFlag) (*wire.NetAddress, error) {
+	// addr will be a net.TCPAddr when not using a proxy.
+	if tcpAddr, ok := addr.(*net.TCPAddr); ok {
+		ip := tcpAddr.IP
+		port := uint16(tcpAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// addr will be a socks.ProxiedAddr when using a proxy.
+	if proxiedAddr, ok := addr.(*socks.ProxiedAddr); ok {
+		ip := net.ParseIP(proxiedAddr.Host)
+		if ip == nil {
+			ip = net.ParseIP("0.0.0.0")
+		}
+		port := uint16(proxiedAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// For the most part, addr should be one of the two above cases, but
+	// to be safe, fall back to trying to parse the information from the
+	// address string as a last resort.
+	host, portStr, err := net.SplitHostPort(addr.String())
+	if err != nil {
+		return nil, err
+	}
+	ip := net.ParseIP(host)
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, err
+	}
+	na := wire.NewNetAddressIPPort(ip, uint16(port), services)
+	return na, nil
+}
+
+// outMsg is used to house a message to be sent along with a channel to signal
+// when the message has been sent (or won't be sent due to things such as
+// shutdown)
+type outMsg struct {
+	msg      wire.Message
+	doneChan chan struct{}
+}
+
+// stats is the collection of stats related to a peer
+type stats struct {
+	statsMtx           sync.Mutex // protects all statistics below here.
+	versionKnown       bool
+	protocolVersion    uint32
+	versionSent        bool
+	verAckReceived     bool
+	timeOffset         int64
+	timeConnected      time.Time
+	lastSend           time.Time
+	lastRecv           time.Time
+	bytesReceived      uint64
+	bytesSent          uint64
+	startingHeight     int32
+	lastBlock          int32
+	lastAnnouncedBlock *wire.ShaHash
+	lastPingNonce      uint64    // Set to nonce if we have a pending ping.
+	lastPingTime       time.Time // Time we sent last ping.
+	lastPingMicros     int64     // Time for last ping to return.
+}
+
+// ShaFunc is a function which returns a block sha, height and error
+// It is used as a callback to get newest block details.
+type ShaFunc func() (sha *wire.ShaHash, height int32, err error)
+
+// Peer provides a basic bitcoin peer for handling bitcoin communications.  The
+// overall data flow is split into 3 goroutines.  Inbound messages are read via
```

Comment:
More reference to block manager.  Also it is making a lot of references to internal function names.  This was fine before when it was not exported, but now this is a public comment, it needs to be changed accordingly.  That is what I meant previously by scrubbing it.  The information about the internal working should be split into a separate NOTE just above it since it is useful information for people digging into the peer package, but it doesn't belong in the exported comment.

Here is what I suggest:

```Go
// NOTE: The overall data flow of a peer is split into 3 goroutines.  Inbound
// messages are read via the inHandler goroutine and generally dispatched to
// their own handler.  For inbound data-related messages such as blocks,
// transactions, and inventory, the data is handled by the corresponding
// message handlers.  The data flow for outbound messages is split into 2
// goroutines, queueHandler and outHandler.  The first, queueHandler, is used
// as a way for external entities to queue messages, by way of the QueueMessage
// function, quickly regardless of whether the peer is currently sending or not.
// It acts as the traffic cop between the external world and the actual
// goroutine which writes to the network socket.

// Peer provides a basic concurrent safe bitcoin peer for handling bitcoin
// communications via the peer-to-peer protocol.  It provides full duplex
// reading and writing, automatic handling of the initial handshake process,
// querying of usage statistics and other information about the remote peer such
// as its address, user agent, and protocol version, output message queueing,
// inventory trickling, and the ability to dynamically register and unregister
// callbacks for handling bitcoin protocol messages.
//
// Outbound messages are typically queued via QueueMessage or QueueInventory. 
// QueueMessage is intended for all messages, including responses to data such
// as blocks and transactions.  QueueInventory, on the other hand, is only
// intended for relaying inventory as it employs a trickling mechanism to batch
// the inventory together.  However, some helper functions for pushing messages
// of specific types that typically require common special handling are
// provided as a convenience.
//
// The Add<Foo>Listener and Remove<Foo>Listener family of functions provide the
// ability for the caller to add and remove handlers, respectively, for each of
// the protocol messages the caller is interested in.
```

Notice the blank uncommented line in between the internal package note and the Peer comment to make sure the internal note does not appear in the doc for Peer.